### ISSUE TITLE
Error When Using S3 Accelerated Transfer Endpoint and No Bucket Name

### ISSUE THREAD

**Issue by davidbitton** [ncw/rclone#3123]
#### What is the problem you are having with rclone?
Errors generated when there is no S3 bucket specified when using a bucket specific accelerated transfer endpoint

#### What is your rclone version (output from `rclone version`)
v1.47.0

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Linux ip-10-0-0-10.ec2.internal 3.10.0-957.10.1.el7.x86_64 #1 SMP Mon Mar 18 15:06:45 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

####  Which cloud storage system are you using? (eg Google Drive)
Dropbox and S3

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
`rclone -vv sync MBLive:/David\ Bitton/ MBLive_S3:/  --stats-one-line --stats-unit bits  -P --stats 2s  --max-backlog 200000`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
```
2019/04/21 02:29:15 DEBUG : rclone: Version "v1.47.0" starting with parameters ["rclone" "-vv" "sync" "MBLive:/David Bitton/" "MBLive_S3:/" "--stats-one-line" "-P" "--stats" "2s" "--max-backlog" "100000"]
2019/04/21 02:29:15 DEBUG : Using config file from "/home/centos/.config/rclone/rclone.conf"
2019/04/21 02:29:16 DEBUG : Dropbox root 'David Bitton': Using root namespace "5362150896"
panic: runtime error: comparing uncomparable type request.ErrInvalidParams

goroutine 158 [running]:
github.com/ncw/rclone/lib/errors.Walk(0x15e1160, 0xc000956a20, 0xc0006c75d0)
        /home/travis/gopath/src/github.com/ncw/rclone/lib/errors/errors.go:65 +0x1a9
github.com/ncw/rclone/lib/pacer.IsRetryAfter(0x15e1160, 0xc000956a20, 0x0, 0x0)
        /home/travis/gopath/src/github.com/ncw/rclone/lib/pacer/pacer.go:256 +0x7e
github.com/ncw/rclone/lib/pacer.(*S3).Calculate(0xc0009b2820, 0x0, 0x0, 0x15e1160, 0xc000956a20, 0xc0001b2400)
        /home/travis/gopath/src/github.com/ncw/rclone/lib/pacer/pacers.go:301 +0x39
github.com/ncw/rclone/fs.(*logCalculator).Calculate(0xc00099e200, 0x0, 0x0, 0x15e1160, 0xc000956a20, 0xc0009ae1e0)
        /home/travis/gopath/src/github.com/ncw/rclone/fs/fs.go:1144 +0x6b
github.com/ncw/rclone/lib/pacer.(*Pacer).endCall(0xc0009ae1e0, 0x0, 0x15e1160, 0xc000956a20)
        /home/travis/gopath/src/github.com/ncw/rclone/lib/pacer/pacer.go:188 +0xa1
github.com/ncw/rclone/lib/pacer.(*Pacer).call(0xc0009ae1e0, 0xc000578640, 0xa, 0x20, 0x123f460)
        /home/travis/gopath/src/github.com/ncw/rclone/lib/pacer/pacer.go:198 +0xc4
github.com/ncw/rclone/lib/pacer.(*Pacer).Call(0xc0009ae1e0, 0xc000578640, 0x40be29, 0xc0006c77e0)
        /home/travis/gopath/src/github.com/ncw/rclone/lib/pacer/pacer.go:216 +0x78
github.com/ncw/rclone/backend/s3.(*Fs).dirExists(0xc00021a160, 0x13f1c80, 0xc00021a298, 0xc0000f4700)
        /home/travis/gopath/src/github.com/ncw/rclone/backend/s3/s3.go:1420 +0xaf
github.com/ncw/rclone/backend/s3.(*Fs).Mkdir(0xc00021a160, 0x0, 0x0, 0x0, 0x0)
        /home/travis/gopath/src/github.com/ncw/rclone/backend/s3/s3.go:1443 +0x3ca
github.com/ncw/rclone/backend/s3.(*Object).Update(0xc0001cb3e0, 0x15debc0, 0xc00093e180, 0x7f284fb5fce0, 0xc00067c320, 0xc0003ba8d0, 0x1, 0x1, 0x40c6b8, 0x10)
        /home/travis/gopath/src/github.com/ncw/rclone/backend/s3/s3.go:1777 +0x6b
github.com/ncw/rclone/backend/s3.(*Fs).Put(0xc00021a160, 0x15debc0, 0xc00093e180, 0x7f284fb5fce0, 0xc00067c320, 0xc0003ba8d0, 0x1, 0x1, 0x40c6b8, 0x60, ...)
        /home/travis/gopath/src/github.com/ncw/rclone/backend/s3/s3.go:1405 +0x101
github.com/ncw/rclone/fs/operations.Copy(0x1614800, 0xc00021a160, 0x0, 0x0, 0xc000ae20ce, 0x21, 0x1613c00, 0xc00067c320, 0xc00011f8a8, 0xc00011f9b8, ...)
        /home/travis/gopath/src/github.com/ncw/rclone/fs/operations/operations.go:317 +0x14c4
github.com/ncw/rclone/fs/sync.(*syncCopyMove).pairCopyOrMove(0xc00098c8c0, 0xc0001d8600, 0x1614800, 0xc00021a160, 0xc00098c998)
        /home/travis/gopath/src/github.com/ncw/rclone/fs/sync/sync.go:295 +0x251
created by github.com/ncw/rclone/fs/sync.(*syncCopyMove).startTransfers
        /home/travis/gopath/src/github.com/ncw/rclone/fs/sync/sync.go:321 +0x9f
c
```

I'm receiving this error when I use an S3 endpoint for an accelerated transfer. The accelerated transfer endpoint includes the bucket name — `bucketname.s3-accelerate.amazonaws.com` — which changes the target use on the command line to `MBLive_S3:` from `MBLive_S3:bucketname`. When I removed the specifically defined endpoint from the config file, I no longer receive this error. According to the S3 [documentation](https://tro.pe/2veiN0B) on Accelerated transfer:
> You can use the accelerate endpoint in the AWS CLI, AWS SDKs, and other tools that  transfer data to and from Amazon S3. If you are using the AWS SDKs, some of the supported languages use an accelerate endpoint client configuration flag so you don't need to explicitly set the endpoint for Transfer Acceleration to bucketname.s3-accelerate.amazonaws.com. For examples of how to use an accelerate endpoint client configuration flag, see [Amazon S3 Transfer Acceleration Examples](https://tro.pe/2veB5ip).
However I don't know how `rclone` is coding against AWS.

**Comment by ncw** [comment]
Can you try this

https://beta.rclone.org/branch/v1.47.0-022-ge275ad80-fix-3123-errors-beta/ (uploaded in 15-30 mins)

It should then print a sensible error!  It won't fix the problem though.

Try including the bucket name in the request `MBLive_S3:bucketname` and setting `--s3-force-path-style=false` and `--s3-endpoint=s3-accelerate.amazonaws.com` - I think that might work.  You can set both of those in the config file too.

**Comment by davidbitton** [comment]
It's in the process of running. Interestingly enough, the first time I ran it, I used the entire accelerated endpoint URL which includes the buckname. it failed, however I see why:
```
centos:rclone-v1.47.0-022-ge275ad80-fix-3123-errors-beta-linux-386/ $ rclone -vv sync MBLive:/David\ Bitton/ MBLive_S3:mblive-s3  --stats-one-line --stats-unit bits  -P --stats 2s  --max-backlog 200000 --s3-force-path-style=false --s3-endpoint=mblive-s3.s3-accelerate.amazonaws.com
2019/04/25 15:55:36 DEBUG : rclone: Version "v1.47.0" starting with parameters ["rclone" "-vv" "sync" "MBLive:/David Bitton/" "MBLive_S3:mblive-s3" "--stats-one-line" "--stats-unit" "bits" "-P" "--stats" "2s" "--max-backlog" "200000" "--s3-force-path-style=false" "--s3-endpoint=mblive-s3.s3-accelerate.amazonaws.com"]
2019/04/25 15:55:36 DEBUG : Using config file from "/home/centos/.config/rclone/rclone.conf"
2019/04/25 15:55:37 DEBUG : Dropbox root 'David Bitton': Using root namespace "5362150896"
2019-04-25 15:56:15 ERROR : : error reading destination directory: RequestError: send request failed
caused by: Get https://mblive-s3.mblive-s3.s3-accelerate.amazonaws.com/?delimiter=%2F&max-keys=1000&prefix=: x509: certificate is valid for *.s3-accelerate.amazonaws.com, s3-accelerate.amazonaws.com, *.s3-accelerate.dualstack.amazonaws.com, s3-accelerate.dualstack.amazonaws.com, *.cf-ofp-us-east-1.s3-accelerate.amazonaws.com, *.cf-ofp-us-east-1a.s3-accelerate.amazonaws.com, *.cf-ofp-us-east-1b.s3-accelerate.amazonaws.com, *.cf-ofp-us-east-1c.s3-accelerate.amazonaws.com, *.cf-ofp-us-east-1d.s3-accelerate.amazonaws.com, *.cf-ofp-us-east-1e.s3-accelerate.amazonaws.com, *.cf-ofp-us-east-1f.s3-accelerate.amazonaws.com, *.cf-ofp-us-west-1.s3-accelerate.amazonaws.com, *.cf-ofp-us-west-1a.s3-accelerate.amazonaws.com, *.cf-ofp-us-west-1c.s3-accelerate.amazonaws.com, *.cf-ofp-ap-northeast-2.s3-accelerate.amazonaws.com, *.cf-ofp-ap-northeast-2a.s3-accelerate.amazonaws.com, *.cf-ofp-ap-northeast-2c.s3-accelerate.amazonaws.com, *.cf-ofp-ap-northeast-1.s3-accelerate.amazonaws.com, *.cf-ofp-ap-northeast-1a.s3-accelerate.amazonaws.com, *.cf-ofp-ap-northeast-1c.s3-accelerate.amazonaws.com, *.cf-ofp-ap-northeast-1d.s3-accelerate.amazonaws.com, *.cf-ofp-sa-east-1.s3-accelerate.amazonaws.com, *.cf-ofp-sa-east-1a.s3-accelerate.amazonaws.com, *.cf-ofp-sa-east-1b.s3-accelerate.amazonaws.com, *.cf-ofp-sa-east-1c.s3-accelerate.amazonaws.com, *.cf-ofp-ap-southeast-1.s3-accelerate.amazonaws.com, *.cf-ofp-ap-southeast-1a.s3-accelerate.amazonaws.com, *.cf-ofp-ap-southeast-1b.s3-accelerate.amazonaws.com, *.cf-ofp-ap-southeast-1c.s3-accelerate.amazonaws.com, *.cf-ofp-ca-central-1.s3-accelerate.amazonaws.com, *.cf-ofp-ca-central-1a.s3-accelerate.amazonaws.com, *.cf-ofp-ca-central-1b.s3-accelerate.amazonaws.com, *.cf-ofp-ap-southeast-2.s3-accelerate.amazonaws.com, *.cf-ofp-ap-southeast-2a.s3-accelerate.amazonaws.com, *.cf-ofp-ap-southeast-2b.s3-accelerate.amazonaws.com, *.cf-ofp-ap-southeast-2c.s3-accelerate.amazonaws.com, *.cf-ofp-us-west-2.s3-accelerate.amazonaws.com, *.cf-ofp-us-west-2a.s3-accelerate.amazonaws.com, *.cf-ofp-us-west-2b.s3-accelerate.amazonaws.com, *.cf-ofp-us-west-2c.s3-accelerate.amazonaws.com, *.cf-ofp-m-us-west-2.s3-accelerate.amazonaws.com, *.cf-ofp-m-us-west-2a.s3-accelerate.amazonaws.com, *.cf-ofp-m-us-west-2b.s3-accelerate.amazonaws.com, *.cf-ofp-m-us-west-2c.s3-accelerate.amazonaws.com, *.cf-ofp-us-east-2.s3-accelerate.amazonaws.com, *.cf-ofp-us-east-2a.s3-accelerate.amazonaws.com, *.cf-ofp-us-east-2b.s3-accelerate.amazonaws.com, *.cf-ofp-us-east-2c.s3-accelerate.amazonaws.com, *.cf-ofp-ap-south-1.s3-accelerate.amazonaws.com, *.cf-ofp-ap-south-1a.s3-accelerate.amazonaws.com, *.cf-ofp-ap-south-1b.s3-accelerate.amazonaws.com, *.cf-ofp-eu-central-1.s3-accelerate.amazonaws.com, *.cf-ofp-eu-central-1a.s3-accelerate.amazonaws.com, *.cf-ofp-eu-central-1b.s3-accelerate.amazonaws.com, *.cf-ofp-eu-central-1c.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-1.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-1a.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-1b.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-1c.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-2.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-2a.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-2b.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-2c.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-3.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-3a.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-3b.s3-accelerate.amazonaws.com, *.cf-ofp-eu-west-3c.s3-accelerate.amazonaws.com, not mblive-s3.mblive-s3.s3-accelerate.amazonaws.com
2019-04-25 15:56:15 INFO  : S3 bucket mblive-s3: Waiting for checks to finish
2019-04-25 15:56:15 INFO  : S3 bucket mblive-s3: Waiting for transfers to finish
2019-04-25 15:56:15 ERROR : S3 bucket mblive-s3: not deleting files as there were IO errors
2019-04-25 15:56:15 ERROR : S3 bucket mblive-s3: not deleting directories as there were IO errors
2019-04-25 15:56:15 ERROR : Attempt 1/3 failed with 2 errors and: not deleting files as there were IO errors
0 / 0 Bytes, -, 0 Bits/s, ETA -^C
```

I went back to your email and dropped the bucket name from the endpoint. How about in lieu of the two aforementioned command line switches, we just add one --s3-accelerated-endpoint?

**Comment by davidbitton** [comment]
As a quick follow-up, if I leave the errors running, eventually it picks back up again. It threw errors again, and then eventually continued one more. However I have no exposure as to files that may have been skipped all together.

**Comment by ncw** [comment]
I merged the fix for the backtrace (which isn't relevant to your issue but is good to fix!)

I'm glad it seems to be working.

> As a quick follow-up, if I leave the errors running, eventually it picks back up again. It threw errors again, and then eventually continued one more. However I have no exposure as to files that may have been skipped all together.

Rclone will retry things that have errors and if there is an error in the total sync, then it will try it again.

If all of those things fail and rclone exits with an error, then just run it again to tidy up.

Things that failed will have ERROR logs.

> I went back to your email and dropped the bucket name from the endpoint. How about in lieu of the two aforementioned command line switches, we just add one --s3-accelerated-endpoint?

It looks like there is configuration for this in the Go SDK - `r.Config.S3UseAccelerate`

Using a command line flag `--s3-use-accelerate-endpoint` or a config variable `use_accelerate_endpoint` would be consistent with the AWS tools.

I added this here - can you give it a go?

https://beta.rclone.org/branch/v1.47.0-024-g9bf148d9-fix-3123-s3-accel-beta/ (uploaded in 15-30 mins)

I stuck it in the advanced config section.  You shouldn't need to set the endpoint or the path style with this flag.

**Comment by davidbitton** [comment]
I can report that I used just the newest command line switch — removed yesterday's new ones — and it's running. There's no DEBUG output telling me what endpoint is being used, but hey, it didn't error, so there's that! Thanks!

**Comment by ncw** [comment]
> I can report that I used just the newest command line switch — removed yesterday's new ones — and it's running. There's no DEBUG output telling me what endpoint is being used, but hey, it didn't error, so there's that! Thanks!

Thanks for testing.  Unfortunately the S3 SDK doesn't surface which endpoint is in use I don't think.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.48
